services/authentication/init_authentication.rb

```diff
@@ -0,0 +1,44 @@
+class AuthenticationService
+  class InitAuthentication < ApplicationService
+    def call
+      init_before_failure
+      create_strategy
+    end
+
+    private
+
+    def init_before_failure
+      Warden::Manager.before_failure do |env, opts|
+        env['REQUEST_METHOD'] = 'POST'
+        env.each do |key, value|
+          env[key]['_method'] = 'post' if key == 'rack.request.form_hash'
+        end
+      end
+    end
+
+    def create_strategy
+      Warden::Strategies.add(:password) do
+        def valid?
+          params.dig('user', 'email') && params.dig('user', 'password')
+        end
+
+        def authenticate!
+          authenticated = AuthenticationService::Authenticate.call(
+            email:    params['user']['email'],
+            password: params['user']['password']
+          )
+
+          if user && authenticated
+            success!(user)
+          else
+            throw(:warden, message: 'Email or password is incorrect.')
```

Comment:
Why `throw` and not `raise` ?

Reply:
@kushniryb well, this is how devise uses warden. 
```ruby
# Deny user access whenever their account is not active yet.
# We need this as hook to validate the user activity on each request
# and in case the user is using other strategies beside Devise ones.
Warden::Manager.after_set_user do |record, warden, options|
  if record && record.respond_to?(:active_for_authentication?) && !record.active_for_authentication?
    scope = options[:scope]
    warden.logout(scope)
    throw :warden, scope: scope, message: record.inactive_message
  end
end
```
I suppose this is because of some under-the-hood warden implementations but I'm not really sure.

Reply:
Ok, FYI - https://hasno.info/ruby-gotchas-and-caveats/